### PR TITLE
Delete Cascade

### PR DESCRIPTION
**Pull request by parsonsmatt**
Before submitting your PR, check that you've:

- [ ] Bumped the version number
- [ ] Documented new APIs with [Haddock markup](https://www.haskell.org/haddock/doc/html/index.html)
- [ ] Added [`@since` declarations](http://haskell-haddock.readthedocs.io/en/latest/markup.html#since) to the Haddock

After submitting your PR:

- [ ] Update the Changelog.md file with a link to your PR
- [ ] Check that CI passes (or if it fails, for reasons unrelated to your change, like CI timeouts)

#909

### CODE REVIEW

File: persistent-test/src/CompositeTest.hs [hunk 1]
```diff
@@ -12,7 +12,11 @@ import Init
 
 
 -- mpsGeneric = False is due to a bug or at least lack of a feature in mkKeyTypeDec TH.hs
-share [mkPersist persistSettings { mpsGeneric = False }, mkMigrate "compositeMigrate", mkDeleteCascade persistSettings { mpsGeneric = False }] [persistLowerCase|
+share
+    [ mkPersist persistSettings { mpsGeneric = False }
+    , mkMigrate "compositeMigrate"
+    , mkDeleteCascade persistSettings { mpsGeneric = False }
```
Comment (parsonsmatt) [inline persistent-test/src/CompositeTest.hs]: Currently this fails with:

```

    /home/matt/Projects/persistent/persistent-test/src/CompositeTest.hs:15:1: error:
        • Couldn't match type ‘Key TestParent’ with ‘[Char]’
            arising from a use of ‘TestChildName’
        • In the first argument of ‘Filter’, namely ‘TestChildName’
          In the expression:
            ((Filter TestChildName) (FilterValue key_aGp3)) Eq
          In the first argument of ‘deleteCascadeWhere’, namely
            ‘[((Filter TestChildName) (FilterValue key_aGp3)) Eq,
              ((Filter TestChildName2) (FilterValue key_aGp3)) Eq,
              ((Filter TestChildAge) (FilterValue key_aGp3)) Eq]’
       |
    15 | share
       | ^^^^^...
```

Previously, we weren't generating *anything* here. But now we're generating something like:

```
deleteWhere [TestChildName ==. k, TestChildName2 ==. k, TestChildAge ==. k]
```

which is obviously not right. So the "fix" in thsi PR is borked.

File: persistent-test/src/DeleteCascadeTest.hs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        liftIO $ void xs `shouldBe` []
+
+    it "should delete by non-primary key refernece" $ do
+        pendingWith "Depends on specifying non-PK references: see issue #909"
+        runDb $ do
+            pk <-insert (PersonX "username0" "name0" Nothing)
+            ak <- insert (AliasX "name0" "othername")
+            deleteCascade pk
+            ma <- get ak
+            liftIO $ void ma `shouldBe` Nothing
```
Comment (parsonsmatt) [inline persistent-test/src/DeleteCascadeTest.hs]: @rvl This should take care of your test case, right?

Comment (rvl) [reply persistent-test/src/DeleteCascadeTest.hs]: @parsonsmatt Thanks for looking at this and trying the fix. Yes it does match my use case. I can work around it by adding more `deleteWhere` statements though. If `persistent-sqlite` could add foreign key constraints to the table, that would be ideal, although I sense there might be some limitation of persistent and SQLite together that prevents this.

Comment (parsonsmatt) [reply persistent-test/src/DeleteCascadeTest.hs]: I think that `persistent-sqlite` does add foreign key indexes. It doesn't specify the cascade mode though. Typically you'd write something like:

```sql
CREATE TABLE foo (
  blah_id INT REFERENCES blah(id) ON DELETE CASCADE
);
```

and when the relevant `blah` gets deleted, the `foo` entries get deleted as well.

I do want to extend the QQ syntax with a cascade-mode for deletes and updates. Something like:

```
Foo
   blahId  BlahId OnDelete Cascade OnUpdate Restrict
```

would be reflected in the generated SQL migrations and then you wouldn't need the "manual" cascading entries.

**Pull request closed without merging**